### PR TITLE
Fixes #36822 - Design new hosts overview page

### PR DESCRIPTION
**Pull request by parthaa**
Update the TableIndexPage as a part of this action to accept things like select all

Works with https://github.com/Katello/katello/pull/10771

### CODE REVIEW

File: config/routes.rb
Comment (jeremylenz) [inline config/routes.rb]: This should be `/hosts/new` to match the other new host routes (`/hosts/new/:id`)

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/Table/Table.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/Table/Table.js:132]: Not sure if this is where the fix is, but the host names need to be links to the host details page :)

**Review by jeremylenz — COMMENTED** [review]
It's finally here! :partying_face: 

A couple things I noticed to start:

- Search works, but autocomplete doesn't
- Kebab is too far to the right

File: webpack/assets/javascripts/react_app/components/HostsIndex/index.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/HostsIndex/index.js]: This should actually be in Katello since it's linking to a Katello page.

Comment (jeremylenz) [reply webpack/assets/javascripts/react_app/components/HostsIndex/index.js]: Should set this up similar to the way it's already done for host details.

See -> https://github.com/Katello/katello/blob/a57828483396b6e02e4b6d849202a0253abe9004/webpack/components/extensions/HostDetails/ActionsBar/index.js#L71

Exported from Katello to Foreman here -> https://github.com/Katello/katello/blob/a57828483396b6e02e4b6d849202a0253abe9004/webpack/global_index.js#L70

Consumed here in Foreman -> 
https://github.com/theforeman/foreman/blob/d329c933388647d3260684d0e3d9bbc9a34397b4/webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/Selectors.js#L6

which is then rendered here -> https://github.com/theforeman/foreman/blob/d329c933388647d3260684d0e3d9bbc9a34397b4/webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/index.js#L49

File: config/routes.rb
Comment (parthaa) [reply config/routes.rb]: Doesnt `hosts/new` point to create hosts ?

Comment (jeremylenz) [reply config/routes.rb]: good point.. Ok, leave as-is then :smile:

**Comment by jeremylenz** [comment]
Also: Should add a setting "Use new All Hosts page" or some such, which will control whether the All Hosts link in the main nav goes to the new or the old page.

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/Table/Table.js
Comment (parthaa) [reply webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/Table/Table.js:132]: Fixed!

**Comment by parthaa** [comment]
> Also: Should add a setting "Use new All Hosts page" or some such, which will control whether the All Hosts link in the main nav goes to the new or the old page.

Updated the menu

File: app/registries/foreman/settings/general.rb
Comment (jeremylenz) [inline app/registries/foreman/settings/general.rb]:
```suggestion
      description: N_("Whether or not to show the new overview page for All Hosts (requires reload of page)"),
```

File: app/registries/foreman/settings/general.rb
Comment (jeremylenz) [inline app/registries/foreman/settings/general.rb]:
```suggestion
      full_name: N_('Show New Host Overview Page'))
```

File: webpack/assets/javascripts/react_app/components/HostsIndex/index.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/HostsIndex/index.js]: This returns JSX, so it should be styled as a React component.

```suggestion
  const RowSelectNode = props => (
    <Td
      select={{
        rowIndex: props.id,
        onSelect: (_event, isSelecting) => {
          selectOne(isSelecting, props.id);
        },
        isSelected: isSelected(props.id),
        disable: false,
      }}
    />
  );
```

Can also be moved to its own file and imported.

In addition, don't use the presence of this component to decide things in your other component. Instead, pass a regular boolean prop down to TableIndexPage, and name it accordingly (`showCheckboxes`, perhaps?)

File: webpack/assets/javascripts/react_app/components/HostsIndex/index.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/HostsIndex/index.js]: This doesn't need to be a function

```suggestion
  const dropdownHref = selectedCount > 0 ? `/change_host_content_source?search=${fetchBulkParams()}` : '';
```

File: webpack/assets/javascripts/react_app/components/HostsIndex/index.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/HostsIndex/index.js]:
```suggestion
      href={dropdownHref}
```

File: webpack/assets/javascripts/react_app/components/HostsIndex/index.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/HostsIndex/index.js]: :partying_face: :rocket:

File: webpack/assets/javascripts/react_app/components/HostsIndex/index.js
Comment (parthaa) [reply webpack/assets/javascripts/react_app/components/HostsIndex/index.js]: Fixed!

File: webpack/assets/javascripts/react_app/components/HostsIndex/index.js
Comment (parthaa) [reply webpack/assets/javascripts/react_app/components/HostsIndex/index.js]: Fixed!

File: webpack/assets/javascripts/react_app/routes/Models/ModelsPage/index.js
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/routes/Models/ModelsPage/index.js]: I didnt check the rest of the PR, but can we not move this out of the table? the reason for the table component is so other components will not handle api calls.
Also this change breaks webhooks which also use the table component (https://github.com/theforeman/foreman_webhooks/blob/master/webpack/ForemanWebhooks/Routes/Webhooks/WebhooksIndexPage/WebhooksIndexPage.js)

Comment (jeremylenz) [reply webpack/assets/javascripts/react_app/routes/Models/ModelsPage/index.js]: It made a lot more sense to me and @parthaa to move it up one level. This way the host-specifc parts (including the host API call) are in the host table component, while the generic table parts are in the table component. Not having access to `response` in the hosts table caused a lot of problems.

Perhaps we could leave it as is and update webhooks instead?

Comment (MariaAga) [reply webpack/assets/javascripts/react_app/routes/Models/ModelsPage/index.js]: Cant the code use redux selectors to access the `response`?

Comment (jeremylenz) [reply webpack/assets/javascripts/react_app/routes/Models/ModelsPage/index.js]: That's a possibility, just seemed a bit less elegant to me.

Personally I feel like needing to access `response` thru Redux defeats the whole purpose of useAPI.

@parthaa thoughts?

Comment (parthaa) [reply webpack/assets/javascripts/react_app/routes/Models/ModelsPage/index.js]: Fixed it in the latest commit. Secret sauce here is this line
```javascript
let response = useAPI(
    replacementResponse ? null : 'get',
    apiUrl.includes('include_permissions')
      ? apiUrl
      : `${apiUrl}?include_permissions=true`,
    {
      ...apiOptions,
      params: defaultParams,
    }
  );
```

Comment (parthaa) [reply webpack/assets/javascripts/react_app/routes/Models/ModelsPage/index.js]: method being null implies that useAPI effects wont do anything.

File: app/registries/foreman/settings/general.rb
Comment (jeremylenz) [inline app/registries/foreman/settings/general.rb]:
```suggestion
    setting('new_hosts_page',
```

This setting only affects one page (All Hosts), so it might be confusing to name it `pages`.

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/Table/SelectAllCheckbox/index.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/Table/SelectAllCheckbox/index.js]:
```suggestion
  // None are selected -> click -> all are selected, or page is selected (depends on canSelectAll)
```

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/Table/TableHooks.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/Table/TableHooks.js]: Saw this on the internet somewhere. [Symbols](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) are always unique values, so this may be a bit more elegant :)

```suggestion
  const forceRender = () => setToggle(Symbol());
```

**Review by jeremylenz — COMMENTED** [review]
some silly nits

File: webpack/assets/javascripts/react_app/components/HostsIndex/index.js
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/HostsIndex/index.js:114]: Why is `creatable` false?

Comment (parthaa) [reply webpack/assets/javascripts/react_app/components/HostsIndex/index.js:114]: I think on the Hosts Index page we need create button as the same level as Hosts header. So decided to add that to the HostIndex component and not use TableIndex for that.

File: webpack/assets/javascripts/react_app/components/HostsIndex/index.js
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/HostsIndex/index.js:112]: Hosts API doesnt pass `can_delete` for each result, so the delete button is always disabled

File: webpack/assets/javascripts/react_app/components/HostsIndex/index.js
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/HostsIndex/index.js:114]: Will that be added in a different pr?

Comment (jeremylenz) [reply webpack/assets/javascripts/react_app/components/HostsIndex/index.js:114]: yeah, we'll be adding Delete right after this PR :)

Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/HostsIndex/index.js:114]: And the create? :upside_down_face:

Comment (jeremylenz) [reply webpack/assets/javascripts/react_app/components/HostsIndex/index.js:114]: oops, misread that!

I can add create in the same PR, if partha doesn't add it here

File: webpack/assets/javascripts/react_app/components/HostsIndex/index.js
Comment (parthaa) [reply webpack/assets/javascripts/react_app/components/HostsIndex/index.js:112]: Fixed

File: app/helpers/application_helper.rb
Comment (jeremylenz) [inline app/helpers/application_helper.rb]: I'd put this under User settings, since lab features is there too?

**Review by jeremylenz — COMMENTED** [review]
Few things to wrap up - 

image

1. Can we fix the kebab placement?
2. Can we suppress the "No results" message while the page is loading?
3. 'Hosts' link should respect the setting every place it's used

**Comment by MariaAga** [comment]
@jeremylenz 
> Can we suppress the "No results" message while the page is loading?

why? its showing "no results - loading"

**Comment by jeremylenz** [comment]
> why? its showing "no results - loading"

'No results' is a lie if it's still loading. We don't know the number of results until it loads.

anyway, we can fix this later, I'm not gonna block on this one :)

**Comment by jeremylenz** [comment]
Decided to leave for follow-up PR:

Multi-delete
Loading message (maybe)
Kebab placement

@parthaa can you please squash the commits and fix lint :)

**Review by jeremylenz — DISMISSED** [review]
APJ. Let's get this in and continue on, since this will unblock everything else :)

thanks @parthaa!

**Comment by parthaa** [comment]
> 'No results' is a lie if it's still loading. We don't know the number of results until it loads.
> 
> anyway, we can fix this later, I'm not gonna block on this one :)

That logic is in TableIndexPage anyway. So once we make a decision all the pages using it should get updated appropriately.

**Review by jeremylenz — APPROVED** [review]

**Review by jeremylenz — CHANGES_REQUESTED** [review]
looks like Foreman and integration tests need a look!

**Review by jeremylenz — APPROVED** [review]

**Pull request merged**
